### PR TITLE
Announce deselection when jumping to top/bottom of page

### LINKED ISSUE #5549: Firefox: Selection not announced when jumping to Top/Bottom of Page
STR:

1. Within Browse Mode, press Ctrl + A to select the entire contents of the document

2. With the current selection highlighted, either use Ctrl + Home/End to jump to the top/bottom of the document.

Expected: To stay consistent with the behaviour experienced when simply cursoring around a page with an active selection, NVDA should announce that text has been unselected when either actions are performed.

Actual: No announcement is made.

### PR DESCRIPTION
**Pull request by keyan**
### Link to issue number:

Fixes #5549

### Summary of the issue:

Movement to the top/bottom of page using ctrl+home/end causes deselection of text but this deselection is not announced.

### Description of how this pull request fixes the issue:
This pull-request modifies `caretMovementScriptHelper` so that movement to the top or bottom of the page sets the `isCollapsed` field when there was a prior text selection. This results in announcing the deselection of text.

### Testing strategy:
I first reproduced the bug in question, video: https://www.youtube.com/watch?v=R1362g2nRYE

Then I applied this patch and tested again. In this video the same movement is applied but the text deselection is announced: https://www.youtube.com/watch?v=v5iOY-_augE

### Known issues with pull request:
It was not clear to me how to add unit tests for this case.

### Change log entries:
Bug fixes

### Code Review Checklist:
- [x] Pull Request description is up to date.
- [ ] Unit tests.
- [ ] System (end to end) tests.
- [x] Manual tests.
- [ ] User Documentation.
- [ ] Change log entry.
- [ ] Context sensitive help for GUI changes.

### CODE REVIEW

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
Thanks for your contribution. Honestly, this looks a bit like a workaround to me (i.e. changing the isCollapsed property while the textInfo itself is still collapsed.
I need to look into this more carefully before I give a final review.

**Comment by keyan** [comment]
Hi Leonard, thanks for the review!

Understandable. I'm happy to try another solution if you have some suggestions for a better approach. I meant to just open the discussion with this initial solution.

File: source/cursorManager.py [hunk 1]
```diff
@@ -126,7 +126,14 @@ def _set_selection(self, info):
...
+		# #5549 - Make sure oldInfo is not collapsed when going to top/bottom
+		# of page so that the deselection is announced.
+		oldSelection = self.makeTextInfo(textInfos.POSITION_SELECTION)
+		isJumpToTopOrBottom = posConstant == textInfos.POSITION_FIRST or posConstant == textInfos.POSITION_LAST
+		if not oldSelection.isCollapsed and isJumpToTopOrBottom:
+			oldInfo.isCollapsed = False
+
 		info=oldInfo.copy()
 		info.collapse(end=self.isTextSelectionAnchoredAtStart)
 		if self.isTextSelectionAnchoredAtStart and not oldInfo.isCollapsed:
```
Comment (seanbudd) [inline source/cursorManager.py:139]: Is this going to be affected and should it be? What happens if you jump to the start/end of the page while the text selection is anchored at the start? Could this be included in the system test?

If this shouldn't be affected, maybe we should have `hasSelectionChanged = oldInfo.isCollapsed or (not oldSelection.isCollapsed and isJumpToTopOrBottom)` and use that boolean to check whether to `speakSelectionChange`

**Review by seanbudd — CHANGES_REQUESTED** [review]
Thanks @keyan,

>It was not clear to me how to add unit tests for this case.

A system test for this could be added in chromeTests. It would be good to select some text then test:

- a selection change is announced when moving the caret 
- a selection change is announced when jumping to the bottom of the page
- a selection change is announced when jumping to the top of the page

**Comment by leonardder** [comment]
I think it should be ok to do something like this:

```python
		oldInfo = self.makeTextInfo(posConstant)
		oldSelection = (
			oldInfo
			if posConstant == textInfos.POSITION_SELECTION else
			self.makeTextInfo(textInfos.POSITION_SELECTION)
		)
		...
		if not oldSelection.isCollapsed:
			speech.speakSelectionChange(oldSelection, selection)
```

**Review by michaelDCurran — CHANGES_REQUESTED** [review]
I think that it should be fixed as follows:
* oldInfo should always be initialized as self.makeTextInfo(textInfos.POSITION_SELECTION) rather than using posConstant, as oldInfo is always and only ever used to start from the existing selection/caret and speak the selection difference.
* info should be initialized as oldInfo.copy() if posConstant is textInfos.POSITION_SELECTION (the default) otherwise initialized as self.makeTextInfo(posConstant)

In other words, oldInfo should always represent the original selection/caret. info on the other hand should start at posConstant (selection or otherwise) and possibly move by a number of units.

File: source/cursorManager.py [hunk 2]
```diff
@@ -126,7 +126,14 @@ def _set_selection(self, info):
 		vision.handler.handleCaretMove(self)
 
 	def _caretMovementScriptHelper(self,gesture,unit,direction=None,posConstant=textInfos.POSITION_SELECTION,posUnit=None,posUnitEnd=False,extraDetail=False,handleSymbols=False):
-		oldInfo=self.makeTextInfo(posConstant)
+		oldInfo = self.makeTextInfo(posConstant)
```
Comment (michaelDCurran) [inline source/cursorManager.py:129]:
```suggestion
		oldInfo = self.makeTextInfo(textInfos.POSITION_SELECTION
```

**Comment by lukaszgo1** [comment]
@keyan  Are you planning to work further on this?
